README.rst
Installation
============

Tachyonic Project devstack currently fully supports `CPython <https://www.python.org/downloads/>`__ 3.6, 3,7.

Requirements
------------

Devstack uses docker to run endpoints and services known as projects in Tachonic.

https://www.docker.com

Quick Install
-------------

.. code:: bash

   $ pip3 install devstack

or

.. code:: bash

   $ pip3 install https://github.com/TachyonicProject/devstack/tarball/latest#egg=devstack

Source Code Install
-------------------

Tachyonic Project devstack infrastructure and code is hosted on `GitHub <https://github.com/TachyonicProject/devstack>`_.
Making the code easy to browse, download, fork, etc. Pull requests are always welcome!

Clone the project like this:

.. code:: bash

    $ git clone https://github.com/TachyonicProject/devstack.git

Once you have cloned the repo or downloaded a tarball from GitHub, you
can install Tachyon like this:

.. code:: bash

    $ cd devstack
    $ pip3 install .

Or, if you want to edit the code, first fork the main repo, clone the fork
to your development area, and then run the following to install it using
symbolic linking, so that when you change your code, the changes will be
automatically available to your app without having to reinstall the package.

**Keep in mind Cythonized C code for all modules is included during install
or develop. During development code alterations you need run setup.py after
each change.**

.. code:: bash

    $ cd devstack
    $ python3 setup.py develop

During development without explicitly running **setup.py** after each change,
you can use the following. It clears the compiled .cpython.so modules to ensure
code is loaded from Python sources.

.. code:: bash

    $ python3 setup.py clean

Usage
-----

Devstack creates multiple docker insances for micro-services that share a common directory on the host machine.

The common directory contains Tachyonic Project Repositories for micro-services and dependencies.
There is an additional directory known as www that contains the project deployment environment.

Example startup:

.. code:: bash

   $ mkdir tachyonic
   $ devstack -s tachyonic

Once this is completed, you can see view wether the docker instances are running using:

.. code:: bash
   
   $ docker ps

Provide at least 2 minutes for all nodes and services to be runnining especially during initial startup.

You can access the Photonic UI via http://localhost:9000

All other endpoints/services are exposed to the host.

+-------------------------+------+----------------+
| Mysql                   | 3306 |                |
+-------------------------+------+----------------+
| UI/Photonic             | 9000 | (RestAPI/JSON) |
+-------------------------+------+----------------+
| Identity/Infinitystone  | 9001 | (RestAPI/JSON) |
+-------------------------+------+----------------+
| Radius/Tradius          | 9002 | (RestAPI/JSON) |
+-------------------------+------+----------------+

To control which modules are started, specify a json file (via ``-m`` switch) with the required builds and ports.
For example, create a file called ``modules.json``:

.. code:: json

    {
        "builds" : ["infinitystone",
                    "tradius",
                    "netrino",
                    "topenstack",
                    "photonic"
                   ],

        "ports" : {"photonic": { "80/tcp": 9000 },
                  "infinitystone": { "80/tcp": 9001 },
                  "tradius": {"80/tcp": 9002,
                         "1812/udp": 1812,
                         "1813/udp": 1813,
                         "1812/tcp": 1812,
                         "1813/tcp": 1813
                        },
                  "netrino": { "80/tcp": 9004 },
                  "topenstack": { "80/tcp": 9005 }
            }
    }

and start with:

.. code:: bash

   $ devstack -s tachyonic -m modules.json


Development
-----------
By default we do not allow to push to our repositories directly.

When editing code ensure you rename origin and add your own fork as origin.

Example:

.. code:: bash

   $ git remote rename origin upstream
   $ git remote add origin git@github.com:cfrademan/tradius.git
   $ git push -u origin development

To restart endpoint gunicorn wsgi applications:

.. code:: bash

   $ devstack -r

When the tachyonic containers start for the first time, their entrypoint scripts does a ``pip install`` on the package
to install the package as well as their dependencies. Because this takes a bit of time, it creates a
``/installed`` file in the container, and only performs this installation if the ``/installed`` file is not present. If you
have the requirement to force a re-installation, simply remove that file from the running container.
For example, for photonic:

.. code:: bash

   $ docker exec photonic rm /installed

Next time when ``devstack -s path`` is run, the package and it's dependacies will be re-installed.


RELEASES.rst
Releases
########

1.0.0 BETA RELEASE (2019/02/06)
-------------------------------

   * Initial Beta Release.
   * Update packages, tradius -> subscriber.


devstack/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (c) 2018-2019 Christiaan Frans Rademan, Dave Kruger.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice, this
#   list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
#
# * Neither the name of the copyright holders nor the names of its
#   contributors may be used to endorse or promote products derived from
#   this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
# THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import argparse
import subprocess
import json
from tempfile import TemporaryFile


from luxon.utils.pkg import Module
from luxon.utils.files import mkdir, Open, chmod, exists, ls, rm, joinpath
from luxon.core.config import Config
from luxon.utils.dk import (start,
                            stop,
                            restart,
                            build,
                            remove_image,
                            remove_container,
                            exec)

from devstack import metadata

builds = ['infinitystone',
          'subscriber',
          'netrino',
          'photonic',
          ]

depends = ['luxon',
           'psychokinetic'
           ]

ports = {'photonic': {'80/tcp': 9000},
         'infinitystone': {'80/tcp': 9001},
         'subscriber': {'80/tcp': 9002,
                     '1812/udp': 1812,
                     '1813/udp': 1813,
                     '1812/tcp': 1812,
                     '1813/tcp': 1813,
                     },
         'netrino': {'80/tcp': 9003},
         }


def execute(*args):
    loginfo = TemporaryFile()
    env = os.environ.copy()
    if '__PYVENV_LAUNCHER__' in env:
        del env['__PYVENV_LAUNCHER__']

    try:
        subprocess.run(*args, stdout=loginfo,
                       stderr=loginfo,
                       check=True, env=env)
        loginfo.seek(0)
        return loginfo.read().decode('utf-8')
    except Exception:
        loginfo.seek(0)
        print(loginfo.read().decode('utf-8'))
        raise

def kill(args):
    for b in builds:
        try:
            print("Stopping %s" % b)
            stop(b)
        except:
            pass
    try:
        print("Stopping sql")
        stop('sql')
    except:
        pass

    try:
        print("Stopping rabbitmq")
        stop('rabbitmq')
    except:
        pass

    try:
        print("Stopping redis")
        stop('redis')
    except:
        pass

def clear(args):
    for b in builds:
        try:
            print("Deleting container %s" % b)
            remove_container(b)
        except:
            pass
    try:
        print("Deleting container sql")
        remove_container('sql')
    except:
        pass

    try:
        print("Deleting container rabbitmq")
        remove_container('rabbitmq')
    except:
        pass

    try:
        print("Deleting container redis")
        remove_container('redis')
    except:
        pass

def delete(args):
    for b in builds:
        try:
            print("Deleting image %s" % b)
            remove_image(b)
        except:
            pass
    try:
        print("Deleting image sql")
        remove_image('sql')
    except:
        pass

    try:
        print("Deleting image rabbitmq")
        remove_image('rabbitmq')
    except:
        pass

    try:
        print("Deleting image redis")
        remove_image('redis')
    except:
        pass

def build_images():
    module = Module('devstack')
    for b in builds:
        print("Building %s" % b)
        docker_file = '/resources/%s/Dockerfile' % b
        docker_obj = module.file(docker_file)
        image, log = build(b,
                           docker_obj)
        for l in log:
            print(l)

def clone_repos():
    for b in builds + depends:
        if not exists(b):
            origin = "https://github.com/TachyonicProject/%s.git" % b
            print(execute(['git', 'clone', '-b', 'development',  origin]))

def reload(args):
    for b in builds:
        print("Reloading %s" % b)
        try:
            exec(b, 'pkill -HUP -f gunicorn3')
        except:
            print("Failed Reloading %s" % b)

def start_env(path):
    clone_repos()
    mkdir('%s/www' % path)
    print("Starting sql")
    start('sql', 'mariadb:latest',
	  ports={'3306/tcp': 3306},
	  MYSQL_ROOT_PASSWORD='tachyonic',
	  MYSQL_DATABASE='tachyonic',
	  MYSQL_USER='tachyonic',
	  MYSQL_PASSWORD='tachyonic')
    print("Starting rabbitmq")
    start('rabbitmq', 'rabbitmq:latest')
    print("Starting redis")
    start('redis', 'redis:latest')

    build_images()
    links = []
    module = Module('devstack')
    for b in builds:
        print("Starting %s" % b)
        module.copy('resources/%s/%s.sh' % (b,b,),
                    path)
        module.copy('resources/%s/%s.nginx' % (b,b,),
                    path)
        start(b, b,
              links=['sql', 'rabbitmq', 'redis',] + links,
              volumes={path: '/opt/tachyonic'},
              ports=ports.get(b),
              hostname=b)
        links.append(b)

def main(argv):
    global builds
    global ports

    description = metadata.description + ' ' + metadata.version
    print("%s\n" % description)

    parser = argparse.ArgumentParser(description=description)
    group = parser.add_mutually_exclusive_group(required=True)

    group.add_argument('-s',
                       dest='path',
                       help='Start/Build environment in path')

    group.add_argument('-c',
                       action='append_const',
                       dest='funcs',
                       const=clear,
                       help='Clear containers in docker environment')

    group.add_argument('-d',
                       action='append_const',
                       dest='funcs',
                       const=delete,
                       help='Clear images in docker environment')

    group.add_argument('-k',
                       action='append_const',
                       dest='funcs',
                       const=kill,
                       help='Shutdown all containers in docker environment')

    group.add_argument('-r',
                       action='append_const',
                       dest='funcs',
                       const=reload,
                       help='Reload Gunicorn Applications in all containers')

    parser.add_argument('-m',
                        help='Load modules from json file at '
                            'specified location',
                        dest='json_file',
                        default=None)


    args = parser.parse_args()

    if args.json_file is not None:
        modules = json.load(open(args.json_file))
        builds = modules['builds']
        ports = modules['ports']

    if args.funcs is not None:
        for func in args.funcs:
            func(args)

    if args.path is not None:
        args.path = os.path.abspath(args.path)
        os.chdir(args.path)
        start_env(args.path)

def entry_point():
    """Zero-argument entry point for use with setuptools/distribute."""
    raise SystemExit(main(sys.argv))


if __name__ == '__main__':
    entry_point()
